File: quantix/utils/formatters.py
"""
Formatters utility module for formatting financial data.
"""
from typing import Optional, List
from datetime import datetime


class Formatters:
    """Utility class for formatting financial data."""

    @staticmethod
    def format_currency(value: float, currency: str = "USD", decimals: int = 2) -> str:
        """
        Format a value as currency.

        Args:
            value: Numeric value.
            currency: Currency code.
            decimals: Number of decimal places.

        Returns:
            Formatted currency string.
        """
        symbols = {
            "USD": "$",
            "EUR": "€",
            "GBP": "£",
            "JPY": "¥"
        }
        symbol = symbols.get(currency, currency + " ")
        return f"{symbol}{value:,.{decimals}f}"

    @staticmethod
    def format_percentage(value: float, decimals: int = 2) -> str:
        """
        Format a value as percentage.

        Args:
            value: Numeric value (0.05 = 5%).
            decimals: Number of decimal places.

        Returns:
            Formatted percentage string.
        """
        return f"{value * 100:.{decimals}f}%"

    @staticmethod
    def format_number(value: float, decimals: int = 2) -> str:
        """
        Format a number with thousand separators.

        Args:
            value: Numeric value.
            decimals: Number of decimal places.

        Returns:
            Formatted number string.
        """
        return f"{value:,.{decimals}f}"

    @staticmethod
    def format_date(date: datetime, fmt: str = "%Y-%m-%d") -> str:
        """
        Format a datetime object as string.

        Args:
            date: Datetime object.
            fmt: Date format string.

        Returns:
            Formatted date string.
        """
        return date.strftime(fmt)

    @staticmethod
    def format_large_number(value: float) -> str:
        """
        Format large numbers with suffixes (K, M, B, T).

        Args:
            value: Numeric value.

        Returns:
            Formatted string with suffix.
        """
        suffixes = [
            (1e12, "T"),
            (1e9, "B"),
            (1e6, "M"),
            (1e3, "K")
        ]

        for threshold, suffix in suffixes:
            if abs(value) >= threshold:
                return f"{value / threshold:.2f}{suffix}"

        return f"{value:.2f}"

File: market/utils/formatters.js
/**
 * Formatters utility module for formatting market data.
 */
export class Formatters {
    /**
     * Format a number as currency.
     * @param {number} value - Numeric value
     * @param {string} currency - Currency code
     * @param {string} locale - Locale string
     * @returns {string}
     */
    static formatCurrency(value, currency = 'USD', locale = 'en-US') {
        return new Intl.NumberFormat(locale, {
            style: 'currency',
            currency: currency
        }).format(value);
    }

    /**
     * Format a number as percentage.
     * @param {number} value - Numeric value (0.05 = 5%)
     * @param {number} decimals - Decimal places
     * @returns {string}
     */
    static formatPercent(value, decimals = 2) {
        return `${(value * 100).toFixed(decimals)}%`;
    }

    /**
     * Format a number with thousand separators.
     * @param {number} value - Numeric value
     * @param {number} decimals - Decimal places
     * @returns {string}
     */
    static formatNumber(value, decimals = 2) {
        return new Intl.NumberFormat('en-US', {
            minimumFractionDigits: decimals,
            maximumFractionDigits: decimals
        }).format(value);
    }

    /**
     * Format large numbers with K/M/B suffixes.
     * @param {number} value - Numeric value
     * @returns {string}
     */
    static formatCompact(value) {
        const suffixes = [
            { threshold: 1e12, suffix: 'T' },
            { threshold: 1e9, suffix: 'B' },
            { threshold: 1e6, suffix: 'M' },
            { threshold: 1e3, suffix: 'K' }
        ];

        for (const { threshold, suffix } of suffixes) {
            if (Math.abs(value) >= threshold) {
                return `${(value / threshold).toFixed(2)}${suffix}`;
            }
        }

        return value.toFixed(2);
    }

    /**
     * Format a date object.
     * @param {Date} date - Date object
     * @param {string} format - Format type ('short', 'long', 'time')
     * @returns {string}
     */
    static formatDate(date, format = 'short') {
        const options = {
            short: { year: 'numeric', month: 'short', day: 'numeric' },
            long: { year: 'numeric', month: 'long', day: 'numeric', weekday: 'long' },
            time: { hour: '2-digit', minute: '2-digit', second: '2-digit' }
        };

        return new Intl.DateTimeFormat('en-US', options[format] || options.short).format(date);
    }

    /**
     * Format price change with color indicator.
     * @param {number} change - Price change value
     * @param {number} changePercent - Change percentage
     * @returns {Object}
     */
    static formatPriceChange(change, changePercent) {
        const isPositive = change >= 0;
        return {
            change: `${isPositive ? '+' : ''}${this.formatCurrency(change)}`,
            percent: `${isPositive ? '+' : ''}${this.formatPercent(changePercent)}`,
            color: isPositive ? 'green' : 'red',
            arrow: isPositive ? '▲' : '▼'
        };
    }

    /**
     * Format volume.
     * @param {number} volume - Volume value
     * @returns {string}
     */
    static formatVolume(volume) {
        return this.formatCompact(volume);
    }
}
